python/src/array.cpp

```diff
@@ -770,37 +770,90 @@ void init_array(nb::module_& m) {
+                return false;
+              }
+            }
             return equal(a, to_array(v, a.dtype()));
           },
           "other"_a)
       .def(
           "__lt__",
-          [](const array& a, const ScalarOrArray v) {
+          [](const array& a,
+             const ScalarOrArray v) -> std::variant<array, bool> {
+            auto check_for_obj = std::get_if<nb::object>(&v);
+            if (check_for_obj) {
+              // return false in case of object comparison which is not mlx
+              // array
+              if (!isMlxCoreArray(*check_for_obj)) {
+                return false;
+              }
+            }
             return less(a, to_array(v, a.dtype()));
           },
           "other"_a)
       .def(
           "__le__",
-          [](const array& a, const ScalarOrArray v) {
+          [](const array& a,
+             const ScalarOrArray v) -> std::variant<array, bool> {
+            auto check_for_obj = std::get_if<nb::object>(&v);
+            if (check_for_obj) {
+              // return false in case of object comparison which is not mlx
+              // array
+              if (!isMlxCoreArray(*check_for_obj)) {
+                return false;
+              }
+            }
             return less_equal(a, to_array(v, a.dtype()));
           },
           "other"_a)
       .def(
           "__gt__",
-          [](const array& a, const ScalarOrArray v) {
+          [](const array& a,
+             const ScalarOrArray v) -> std::variant<array, bool> {
+            auto check_for_obj = std::get_if<nb::object>(&v);
+            if (check_for_obj) {
+              // return false in case of object comparison which is not mlx
+              // array
+              if (!isMlxCoreArray(*check_for_obj)) {
+                return false;
+              }
+            }
             return greater(a, to_array(v, a.dtype()));
           },
           "other"_a)
       .def(
           "__ge__",
-          [](const array& a, const ScalarOrArray v) {
+          [](const array& a,
+             const ScalarOrArray v) -> std::variant<array, bool> {
+            auto check_for_obj = std::get_if<nb::object>(&v);
+            if (check_for_obj) {
+              // return false in case of object comparison which is not mlx
+              // array
+              if (!isMlxCoreArray(*check_for_obj)) {
+                return false;
+              }
+            }
             return greater_equal(a, to_array(v, a.dtype()));
           },
           "other"_a)
       .def(
           "__ne__",
-          [](const array& a, const ScalarOrArray v) {
+          [](const array& a,
+             const ScalarOrArray v) -> std::variant<array, bool> {
+            auto check_for_obj = std::get_if<nb::object>(&v);
+            if (check_for_obj) {
+              // return true in case of object inequlity which is not mlx array
+              if (!isMlxCoreArray(*check_for_obj)) {
+                return true;
+              }
```

Comment:
It might be easier with something like `to_array_or_bool` that returns a variant 🤔

Reply:
I think I did not quite understand what you meant. I wrote a function `to_array_or_bool` in `utils.h` as you can see:
```
inline std::variant<array, bool> to_array_or_bool(
    const ScalarOrArray& v,
    std::optional<Dtype> dtype = std::nullopt) {
  // try to convert to array, if it fails, return false
  try {
    return to_array(v, dtype);
  } catch (const std::exception& e) {
    return false;
  }
}
```
Do you mean that we replace this, instead of the check that we are doing with `isMlxCoreArray`?

Reply:
Yes exactly.